### ISSUE TITLE
cygwin DLL address conflict with version 5.41.6

### ISSUE THREAD

**Issue by tonycoz** [Perl/perl5#22695]
Module:

**Description**
Similar to #22104 we're seeing address conflicts loading LangInfo.dll in CI:
```
      0 [main] perl 13868 child_info_fork::abort: address space needed by 'Langinfo.dll' (0x190000) is already occupied
Can't fork, trying again in 5 seconds at lib/MakeMaker/Test/Utils.pm line 325.
      0 [main] perl 13870 child_info_fork::abort: address space needed by 'Langinfo.dll' (0x190000) is already occupied
Can't fork, trying again in 5 seconds at lib/MakeMaker/Test/Utils.pm line 325.
      0 [main] perl 13872 child_info_fork::abort: address space needed by 'Langinfo.dll' (0x190000) is already occupied
Can't fork, trying again in 5 seconds at lib/MakeMaker/Test/Utils.pm line 325.
      0 [main] perl 13983 child_info_fork::abort: address space needed by 'Langinfo.dll' (0x190000) is already occupied
Can't fork, trying again in 5 seconds at lib/MakeMaker/Test/Utils.pm line 325.
```

And rebase shows the conflict:
```
$ rebase -i `find . -name '*.dll'` ./perl.exe | grep -F '*'
/home/tony/dev/perl/git/perl/cygperl5_41_6.dll                          base 0x00059f6d0000 size 0x010d8000 *
/home/tony/dev/perl/git/perl/lib/auto/I18N/Langinfo/Langinfo.dll        base 0x00059fe80000 size 0x00025000 *
```

And can be similarly reproduced:
```
$ ./perl -Ilib -MI18N::Langinfo -efork
      0 [main] perl 16510 child_info_fork::abort: address space needed by 'Langinfo.dll' (0x400000) is already occupied
```

I plan to prepare a fix and perldelta note PR.

**Steps to Reproduce**
1. build perl, I've only tested with `-DDEBUGGING`, there may or may not be a conflict without it
2. run `./perl -Ilib -MI18N::LangInfo -efork`

**Expected behavior**
No error on fork.

**Perl configuration**

```
Summary of my perl5 (revision 5 version 41 subversion 6) configuration:
  Commit id: 161ff7336a85931cf738fbfbffcf68de4d2f1c81
  Platform:
    osname=cygwin
    osvers=3.5.4-1.x86_64
    archname=cygwin-thread-multi
    uname='cygwin_nt-10.0-19045 ganymede 3.5.4-1.x86_64 2024-08-25 16:52 utc x86_64 cygwin '
    config_args='-des -Dusedevel -DDEBUGGING'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='gcc'
    ccflags ='-D_GNU_SOURCE -fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector-strong -D_FORTIFY_SOURCE=2'
    optimize='-O3 -g'
    cppflags='-D_GNU_SOURCE -fwrapv -DDEBUGGING -fno-strict-aliasing -pipe -fstack-protector-strong'
    ccversion=''
    gccversion='12.4.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='g++'
    ldflags =' -Wl,--enable-auto-import -Wl,--export-all-symbols -Wl,--enable-auto-image-base -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/lib /usr/lib/w32api /usr/local/lib /lib
    libs=-lpthread -ldl
    perllibs=-lpthread -ldl
    libc=/usr/lib/libcygwin.a
    so=dll
    useshrplib=true
    libperl=cygperl5_41_6.dll
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=dll
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags=' '
    lddlflags=' --shared  -Wl,--enable-auto-import -Wl,--export-all-symbols -Wl,--enable-auto-image-base -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl):
  Compile-time options:
    DEBUGGING
    HAS_LONG_DOUBLE
    HAS_STRTOLD
    HAS_TIMES
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_HASH_FUNC_SIPHASH13
    PERL_HASH_USE_SBOX32
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_TRACK_MEMPOOL
    PERL_USE_DEVEL
    PERL_USE_SAFE_PUTENV
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
    USE_REENTRANT_API
  Built under cygwin
  Compiled at Oct 23 2024 12:47:36
  @INC:
    lib
    /usr/local/lib/perl5/site_perl/5.41.6/cygwin-thread-multi
    /usr/local/lib/perl5/site_perl/5.41.6
    /usr/local/lib/perl5/5.41.6/cygwin-thread-multi
    /usr/local/lib/perl5/5.41.6
```

**Comment by jkeenan** [comment]
As of Dec 26 2024, we're seeing new test failures on the `cygwin` rig in our GH CI setup.  See, *e.g.,* https://github.com/Perl/perl5/actions/runs/12481530925/job/34915098507?pr=22871.  Is this more of the same problem discussed in this open ticket?

**Comment by haarg** [comment]
That appears to be an separate problem, related to `kill`.

**Comment by tonycoz** [comment]
I suspect [this](https://cygwin.com/pipermail/cygwin/2024-December/256975.html) is the problem which has a cygwin patch available, we just need to wait for a 3.5.6 release, which will probably be delayed due to the holidays.

**Comment by bulk88** [comment]
FWIW, I tried a Cygwin blead perl build today. I can not reproduce this bug ticket's 2 test fails. I'll requote them ```../cpan/autodie/t/kill.t Failed tests:  2, 6```  ```../ext/POSIX/t/wrappers.t  Failed tests:  74, 76```. Whatever the problem in this ticket, , the test fails are super specific to _something_

So the bug is specific to
 \# 1 exactly one version/build number Cygwin platform
 \# 2 exactly one version/build number of Windows 10/11 OS
 \# 3 must have multi phy-CPUs , gaming rigs never count because those r solo CPUs

Debug it without pushing to blead brand a commit with C intrinsic function MS ```DebugBreak();```,  MS ```__debugbreak();``` or GCC ```__builtin_trap();``` . To force the GH cloud VM to execute an intentional SEGV/SIGILL , then somehow over TCPIP, get a full core dump off that runner, or attach GDB/MSVC debugger-over-TCPIP, to the suspended process on the runner VM.

Supposedly GH Actions VMs aren't WWW air-gapped, see these posts from google.

https://discourse.julialang.org/t/non-deterministic-segfault-only-triggered-on-github-actions-tips-for-debugging/53367/3
https://stackoverflow.com/questions/77005/how-to-automatically-generate-a-stacktrace-when-my-program-crashes

This cygwin failure reminds me alot about this bug

 [rt.perl.org#88840]
 Searchable as RT88840$ https://github.com/Perl/perl5/issues/11267
 [rt.perl.org#88840] t/op/fork.t blocks in external process test on Win32

I fixed it in

https://github.com/Perl/perl5/commit/d903973c0527ee5c9f9f559e3e0e3f1aac2ab1cc
fix RT # 88840, don't terminate a child fork psuedo process in DLL Loader Lock

There was also this ticket which at first was ID'ed and fixed as real not-ithread-compatible core interp mem corruption, but the ticket got wrongly tangled up with fork.t race-condition

[rt.perl.org#109718] Searchable as RT109718$ https://github.com/Perl/perl5/issues/11927
fork.t fails on Win32 since v5.15.4-465-g676a678 

but this absurd race-condition was at first git false-bisected to the real and fixed but irrelevant mem corruption bug introduced in

https://github.com/Perl/perl5/commit/676a678ac0683a727a07d56ed1a6e1fb59085d5a
narrower localisation of PL_compcv around eval


``` [rt.perl.org # 88840] t/op/fork.t blocks in external process test on Win32``` is the nastiest race condition I've ever debugged in my life. I ( bulk88 ) am the only person on earth that could reproduce it. No other Win32-using P5P dev could repro it. It was unsolveable. But I could hit repro it, because of a combination of hit-by-lighting super rare reasons. I want to joke it couldn't be repro-ed by any one else, since the race bug in Perl, was  _pure hardware failure_ , not a coding bug. And my system had real _hardware failure_ on its PCBs.

Here are some terms I will use, Kernel == disk file ```ntoskrnl.exe```.  Timeslice == CS PhDs call it ```quantum```.

\# 1 factor ```#88840 t/op/fork.t``` required WinNT Kernel 5.2 (aka Server2003/XP for AMD64). Win 2000 (5.0) and WinXP (5.1) kernels are incapable of repro-ing. ```#88840 t/op/fork.t``` requires a kernel/hal.dll that knows what NUMA is (added 5.2), on a recognized NUMA architecture motherboard. Basically you MUST have a Server 2003 x64, on a AMD HyperTransport bus motherboard, with atleast TWO Physical CPUs, and the DDR-RAM sticks electro-mechanically hang off the CPUs, absolutely no northbridge chip. Vendor, Intel/AMD doesn't matter, its all x64.

I had 4 physical CPUs, in 4U HP Proliant DL585 G1, which had 4 physical Opteron 846 @ 2.0Ghz. 2 cores per physical CPU, total of 8 cores for WinOS. I will bet the NT 5.2 kernel turns off NUMA flag at boot time for all Multi-Core but One Chip X64 motherboards. Nobody has SMP Desktops/Towers >= 2010, not even gamers with unlimited budgets. But I am using Perl in 2013 on an ancient big iron 4U enterprise server WITH Win64 5.2.

\# 2 factor My HP Proliant DL585 had the slowest enterprise SCSI RAID controller ever built in the 21st century. Seek time was a horrible 65-80 milliseconds in RAID 0. I never figured out why. HP Proliant BIOS did not allow booting from a PCI-X card,  Choices were CdRom or HP SmartArray or floppy. No PCI-X cards. so I couldn't boot off my added SATA card. Looking back, I think it was because I got it used, and the HP BIOS yearly license keys were expired, so I was punished in a "shareware mode". The HP ILO hardware module was always nagging for $.

\# 3 Kernel must be doing the thread pinning/process pinning optimization.  I speculate, this is all of course reverse engineering, since you can't git-blame Windows  *cough* *cough*.  My single stepping and reading oldnewthings/reactos repo/Russinovich/Alex Ionescu/conference talks by MS engineers, make guesses pretty accurate.

So ```#88840 t/op/fork.t``` repro-ing is totally dependent on WinNT 5.2 thread/process pin optimization algo happening. Which is where each WinNT Ring 0 C89 struct KTHREAD object ( Ring 3 user mode thread Class),  each KTHREAD is owned by a KPCR object in a linked list. KPCR == [Physical not-logical] Processor Control Region. WinXP 5.1, Win2000 5.0, and before, kept KPCRs in an EXTERN_C global array C storage in file ntoskrnl.exe.  Because of NUMA, or adding NUMA, All DDR chips and all cache lines (4KB pages? IDK) are physically owned by 1 CPU in a group of many. Sending packets thru HyperTransport on the motherboard PCB to access remote DDR sticks is much (???) slower than local sticks. 

So the wonderful NT 5.2 kernel, has a new feature, C function call Sleep(0) (and select()/wait()), is physical processor local only. Actually, every syscall Ring 3->0 is phy-CPU local. MS's new algo is, is cur_exec KTHREAD syscalls to Ring 0, and wants to de-schedule/block, Ring 3 C stack and Ring 3 CPU registers wont be detached from the logical CPU. No context switching!

Instead each CPU core only iterates through its lock-free local linked of KTHREADs looking for READY_TO_RUN flag, If no READY_TO_RUN flags are found, Sleep(0) insta (50 ns?) returns. Even if on another phy-CPU, all COREs are maxed out 100%, and and there is blocking happening and"performance killing" on a deeper and deeper ```->mPendingTimeslotsAvail``` queue.

There is no pool global pool and no mutex/CRITICAL_SECTION anymore. No contention. But also the kernel WONT move a process or KTHREAD away from the "pinned" favorite phy-CPU. Now we have psuedo lock inversion. Atleast until wall time hits the next multiple of 15 milliseconds. MS Public API contract for wall time update granularity and timeslice unit has been 15 mins from 1993 NT 3.1 to my NT 6.1 (7). maybe NT 10.0????

Suddenly the NT3.1-5.1 behavior of CreateProcess() CreateThread() always context switching and descheduling caller Ring3 KTHREAD (parent perl ithread), and an assumed mandatory (thats UB!!!) next 15 ms timeslot on brand new child THREAD/PROCESS obj thru priority boosting algorithm. CHANGED! 

The race bugs can't be debugged in a C debugger, without setting breakpoints in dissaembler view, and using thread freeze/resume'ing BOTH THREADS at different times, and SINGLE STEP one thread at a time, until the "stars" align. and the deadlock/crash/panic happens.

Plus POSIX/Perl are fundamentally unstable/using un-inited variable, if it involves  64bit int ```time_t```. ```time_t a1 = (U64) a2 / 1000;``` is always "data loss", not "precision loss".

8 byte doubles/NVs are worse, 53 bit ints, but perl/win32/*.c  has no NV usage.

https://github.com/Perl/perl5/commit/d903973c0527ee5c9f9f559e3e0e3f1aac2ab1cc

The race condition bug came back recently and was fixed again!!!!

[MS Public API contract for wall time update granularity and timeslice unit.](https://github.com/Perl/perl5/issues/16529)

[win32.c: rework the waitpid(-1, WNOHANG) fix](https://github.com/Perl/perl5/commit/81295a42ad856da581d910c540f5bb7a18b8f7e2)

And in 2012 the race had many parts

[fix over/underflow issues in win32_msgwait](https://github.com/Perl/perl5/commit/001e9f8966c22ce5f6e43f663bffdffb73bf623c)

[Avoid race codition when setting process exit code on Windows.](https://github.com/Perl/perl5/commit/82e24582a585cdcc94ac54b3e77a325e7aa89846)

 I think these Perl, Windows, and "time IPC problems" race problems, can't be repro-ed on any developers in a VM on any work/home machines, and can't be reproduced on a developer's day job's rackmount servers either. VM or no VM. I'll hypothesize the reason is, AWS/Azure/GCP, are always pegged at 100% CPU usage "host" hypervisor wise. ALL 640 cores per rackmount, every rackmount, is 100% CPU usage 24/7. No dev or company will run other peoples workloads on at-home/on-campus hardware. 

Dynamic spot pricing, not-public compute time auctions, etc, datacenters became airline tickets Every second/a server's CPU is idle, is lost money like empty airline seats after takeoffs. Crypto mining, weather data, and malware analytics firms, will always pay for 7 seconds TTFB no-guarantees ```renice -n 20``` VM instances.


System details for my blead CygPerl, that CAN NOT reproduce the 

 https://github.com/Perl/perl5/actions/runs/12481530925/job/34915098507?pr=2287 and passes 

bead @ commit 8f5aa2238548f7e510cad92826b8d433860d2467 - Chad Granum - 12/23/2024 4:58:25 PM - cpan/Test-Simple - Update to version 1.302206
<details>
  <summary>-V perl</summary>
I used Configure -D. ```cygwin1.dll``` calls itself 3004.9.0.0 /  3.4.10. OS Windows 7 Service Pack 1 X64 6.1.7601

```
Owner@LVT420 /cygdrive/c/sources/perl52/t
$ ../perl -I../lib -V
Summary of my perl5 (revision 5 version 41 subversion 8) configuration:

  Platform:
    osname=cygwin
    osvers=3.4.10-1.x86_64
    archname=cygwin-thread-multi
    uname='cygwin_nt-6.1-7601 lvt420 3.4.10-1.x86_64 2023-11-29 12:12 utc x86_64 cygwin '
    config_args='-d -Dusedevel'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='gcc'
    ccflags ='-D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -D_FORTIFY_SOURCE=2'
    optimize='-O3'
    cppflags='-D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong'
    ccversion=''
    gccversion='11.5.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='g++'
    ldflags =' -Wl,--enable-auto-import -Wl,--export-all-symbols -Wl,--enable-auto-image-base -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/lib /usr/lib/w32api /usr/local/lib /lib
    libs=-lpthread -ldl
    perllibs=-lpthread -ldl
    libc=/usr/lib/libcygwin.a
    so=dll
    useshrplib=true
    libperl=cygperl5_41_8.dll
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=dll
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags=' '
    lddlflags=' --shared  -Wl,--enable-auto-import -Wl,--export-all-symbols -Wl,--enable-auto-image-base -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl):
  Compile-time options:
    HAS_LONG_DOUBLE
    HAS_STRTOLD
    HAS_TIMES
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_HASH_FUNC_SIPHASH13
    PERL_HASH_USE_SBOX32
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_DEVEL
    PERL_USE_SAFE_PUTENV
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
    USE_REENTRANT_API
  Built under cygwin
  Compiled at Jan  2 2025 05:34:47
  @INC:
    ../lib
    /usr/local/lib/perl5/site_perl/5.41.8/cygwin-thread-multi
    /usr/local/lib/perl5/site_perl/5.41.8
    /usr/local/lib/perl5/5.41.8/cygwin-thread-multi
    /usr/local/lib/perl5/5.41.8

Owner@LVT420 /cygdrive/c/sources/perl52/t
```
</details>

**Comment by tonycoz** [comment]
The fork error bug is well understood, and had a workaround for CI and workaround documentation in perldelta in #22696, and a more permanent fix in #22853.

You'll only see this error specifically with a perl with version 5.41.6, due to the way automatic base addresses are calculated.

The kill/signal bug is completely unrelated and should have been reported as a new ticket, the symptoms didn't match the original report.  That kill/signal bug requires cygwin 3.5.5, you don't say what you tested against.  I reproduced it on my desktop:
```
tony@GANYMEDE ~/dev/perl/git/perl
$ uname -a
CYGWIN_NT-10.0-19045 GANYMEDE 3.5.5-1.x86_64 2024-12-20 16:19 UTC x86_64 Cygwin

tony@GANYMEDE ~/dev/perl/git/perl
$ git describe
v5.41.7-14-g6c4c496a61

tony@GANYMEDE ~/dev/perl/git/perl
$ ./Configure -des -Dusedevel && make -j4 test TEST_FILES=../ext/POSIX/t/wrappers.t
...
cd t && (rm -f perl.exe; /usr/bin/ln.exe -s ../perl.exe perl.exe)
PATH=/home/tony/dev/perl/git/perl:.:/usr/bin/ ./runtests choose
t/../ext/POSIX/t/wrappers ... #   Failed test 'kill'
#   at t/wrappers.t line 215.
#          got: '0'
#     expected: '1'
#   Failed test 'raise'
#   at t/wrappers.t line 217.
#          got: '0'
#     expected: '1'
# Looks like you failed 2 tests of 87.
FAILED at test 74
Failed 1 test out of 0, 0.00% okay.
        ../ext/POSIX/t/wrappers.t
### Since not all tests were successful, you may want to run some of
...
```
cygwin have a patch for what looks like the signal issue, which I linked above, it's just waiting on people to get back from holidays and catch up.
